File: otto-android/app/src/main/java/to/ottomot/driftd/routebuilder/RouteBuilderUiState.kt
package to.ottomot.driftd.routebuilder

enum class RouteBuilderUiState {
    SET_START,
    SET_FINISH,
    GENERATING_ROUTE,
    ROUTE_READY,
    MANUAL_PLOT,
    EDIT_ROUTE,
}

object RouteBuilderUiStateResolver {
    data class Inputs(
        val hasStart: Boolean,
        val hasFinish: Boolean,
        val isRunningGuidedGeneration: Boolean,
        val isSnapping: Boolean,
        val hasRoutePath: Boolean,
        val checkpointCount: Int,
        val isManualMode: Boolean,
        val isEditMode: Boolean,
        val hasCompletedGuidedGeneration: Boolean,
        val guidedBackToStartStep: Boolean,
    )

    fun resolve(inputs: Inputs): RouteBuilderUiState {
        if (inputs.isRunningGuidedGeneration ||
            (
                inputs.isSnapping &&
                    inputs.hasStart &&
                    inputs.hasFinish &&
                    !inputs.isManualMode &&
                    !inputs.isEditMode &&
                    !inputs.hasCompletedGuidedGeneration
                )
        ) {
            return RouteBuilderUiState.GENERATING_ROUTE
        }

        if (inputs.isEditMode) {
            return RouteBuilderUiState.EDIT_ROUTE
        }

        if (inputs.isManualMode) {
            if (inputs.hasStart && inputs.hasFinish && inputs.hasRoutePath) {
                return RouteBuilderUiState.EDIT_ROUTE
            }
            return RouteBuilderUiState.MANUAL_PLOT
        }

        if (inputs.hasCompletedGuidedGeneration &&
            inputs.hasStart &&
            inputs.hasFinish &&
            inputs.hasRoutePath
        ) {
            return RouteBuilderUiState.ROUTE_READY
        }

        if (inputs.guidedBackToStartStep || !inputs.hasStart) {
            return RouteBuilderUiState.SET_START
        }

        if (!inputs.hasFinish) {
            return RouteBuilderUiState.SET_FINISH
        }

        return RouteBuilderUiState.SET_FINISH
    }
}

File: otto-mobile/otto-mobile/RouteBuilderUIState.swift
import Foundation

enum RouteBuilderUIState: Equatable {
    case setStart
    case setFinish
    case generatingRoute
    case routeReady
    case manualPlot
    case editRoute
}

enum RouteBuilderUIStateResolver {
    struct Inputs: Equatable {
        var hasStart: Bool
        var hasFinish: Bool
        var isRunningGuidedGeneration: Bool
        var isSnapping: Bool
        var hasRoutePath: Bool
        var checkpointCount: Int
        var isManualMode: Bool
        var isEditMode: Bool
        var hasCompletedGuidedGeneration: Bool
        var guidedBackToStartStep: Bool
    }

    static func resolve(_ inputs: Inputs) -> RouteBuilderUIState {
        if inputs.isRunningGuidedGeneration || (inputs.isSnapping && inputs.hasStart && inputs.hasFinish && !inputs.isManualMode && !inputs.isEditMode && !inputs.hasCompletedGuidedGeneration) {
            return .generatingRoute
        }

        if inputs.isEditMode {
            return .editRoute
        }

        if inputs.isManualMode {
            if inputs.hasStart && inputs.hasFinish && inputs.hasRoutePath {
                return .editRoute
            }
            return .manualPlot
        }

        if inputs.hasCompletedGuidedGeneration && inputs.hasStart && inputs.hasFinish && inputs.hasRoutePath {
            return .routeReady
        }

        if inputs.guidedBackToStartStep || !inputs.hasStart {
            return .setStart
        }

        if !inputs.hasFinish {
            return .setFinish
        }

        return .setFinish
    }
}
